### PR TITLE
Feature: Shows wallet balances on the UI

### PR DESCRIPTION
**Pull request by harishmaddali**
Due to restrictions on calling the Paxos Dev APIs from the browser, I have created our own NodeJS backend to handle the Paxos Dev APIs. You can find the code here: https://github.com/harishmaddali/paxos_trader_be

**Architecture:**
1. The react code takes care of obtaining the OAuth2 Token. `PAXOS_CLIENT_ID` and `PAXOS_CLIENT_SECRET` values are needed in the `.env` file to obtain the token.
2. The auth token is sent as a query parameter to our NodeJS backend endpoints.
3. NodeJS APIs read the auth token from the request and makes the PAXOS v2 API calls to fetch the data (In our case, Wallet Balance).
4. Wallet Balance response from the NodeJS API is shown on the react UI.

**Steps to setup the backend:**
1. Clone the repository: https://github.com/harishmaddali/paxos_trader_be
2. Navigate into the code repository, Run `npm install` to install the required packages.
3. Run `node server.js` to run the server.

**Future Scope:**
1. NodeJS endpoints can be extended to handle all the Paxos v2 API calls
2. React Application could be extended to handle the authorization of any user.
3. Refreshing of the token can be handled on the react application to always have the latest token.

**Challenges faced during the development:**
1. Paxos API management page seemed to be supporting only V2 APIs. Faced challenges to make the V1 code working. So, I have decided to use the V2 APIs.
2. Paxos APIs blocked the preflight requests from the browser, so I spent significant time trying to find the solution. Eventually figured out that we should have our own backend code that manages PAXOS API calls.
3. Paxos team took a long time to accept my application which ate up a significant time.

Leaving the challenges aside, finally, I was able to get the APIs working and was able to show the wallet balances on the UI.

Waiting to hear your thoughts about the approach and the code.

**Note:** I still have to transfer the `paxos_trader_be` repo to your account. I will transfer it soon.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,134 +1,148 @@
...
-    const itBitHost = 'https://api.itbit.com/v1/';
-    axios.get(itBitHost + `markets/${currency}/ticker`)
-      .then(function (response) {
-        const bid = Number(response.data.bid).toFixed(2);
-        const displayBid = !isNaN(bid) ? bid : '';
-        const when = response.data.serverTimeUTC;
-        if (currency === 'XBTUSD') setCurrentBTCprice({ 'bid': displayBid, 'when': when });
-        if (currency === 'ETHUSD') setCurrentETHprice({ 'bid': displayBid, 'when': when });
-        if (currency === 'LTCUSD') setCurrentLTCprice({ 'bid': displayBid, 'when': when });
+  useEffect(() => {
```
Comment (durrantm) [inline src/App.js:16]: Is there a reason for having two useEffect ? Could they be combined with no downside ?

File: .gitignore [hunk 2]
```diff
@@ -21,3 +21,6 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
+
+# Environment
+.env
```
Comment (durrantm) [inline .gitignore:26]: i usually keep this

File: src/App.js [hunk 3]
```diff
@@ -151,18 +168,20 @@ function App() {
...
         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
         {Date(Date.now()).substring(16, 24)}
         {theHours}
       </div>
-    </div >
+    </div>
   );
 }
-export default App;
+export default App;
```
Comment (durrantm) [inline src/App.js:187]: not needed by src/index.js ?

File: src/App.js (hunk 1 above)
Comment (harishmaddali) [reply src/App.js:16]: I wanted to have the code structured in a better way. So, I added the balances related calls on one useEffect and the market data related call on another useEffect.
In the future, if you decided to divide the application into multiple components, you can just copy each use effect into respective components.

Both the useEffects can be combined into one, but the code readability gets affected because of that. For better readability, it is better to have multiple useEffects.

File: src/App.js (hunk 3 above)
Comment (harishmaddali) [reply src/App.js:187]: @durrantm I didn't get the question you were asking. Can you rephrase it for me?

File: src/App.js [hunk 4]
```diff
@@ -1,134 +1,148 @@
-import React, { useState, useEffect, useCallback } from 'react';
-import './App.css';
-import axios from 'axios';
-import { sha256 } from 'sha256';
+import React, { useState, useEffect, useCallback } from "react";
+import "./App.css";
+import axios from "axios";
+import { sha256 } from "sha256";
```
Comment (durrantm) [inline src/App.js:4]: This ended up uneeded so i removed it.
pls confirm it is unneeded.

File: src/App.js [hunk 5]
```diff
@@ -1,134 +1,148 @@
...
-        if (currency === 'LTCUSD') setCurrentLTCprice({ 'bid': displayBid, 'when': when });
+  useEffect(() => {
+    getToken();
+  }, []);
+
+  useEffect(() => {
+    if (token) {
+      getWallet();
+    }
+  }, [token]);
```
Comment (durrantm) [inline src/App.js:24]: This works but I do have one question here:  My linter rules say "  Line 24:6:  React Hook useEffect has a missing dependency: 'getWallet'. Either include it or remove the dependency array".  That is a message I ran into myself in other cases and apps.  Doing either of the things suggested leads to more errors and in some cases (like removing it) to infinite networks calls.  I got burned by that before myself :)  This code is doing ok however, no infinite calls.  Do you know of a change to safely (!) remove the warning?  Ultimately I can just eslint ignore the specific warning for that line but I wanted to check with you to see if you knew or had an opinion.  Thanks Harish !

File: src/App.js (hunk 4 above)
Comment (harishmaddali) [reply src/App.js:4]: Yeah. It is not needed.

File: src/App.js (hunk 5 above)
Comment (harishmaddali) [reply src/App.js:24]: I just read about that lint warning. To fix it, all we have to do is move the getWallet function declaration into the useEffect. That should resolve the issue.

Here is a good StackOverflow solution for reference. They properly explained this.
https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook
